File: src/regulae/training.py
"""Training orchestration for the learned model.

The pairwise training pipeline runs several stages in sequence:

1. **Segment-level EM.** Iterates alignment and segment-table
   updates until the total corpus cost converges.
2. **Displacement aggregation.** One pass over the converged
   alignments to build the feature displacement distribution.
3. **Context conditioning.** A greedy BIC-gated split loop
   looks for immediate-neighbour conditioning environments that
   improve the fit; committed splits become conditioned
   correspondences.
4. **Chunk promotion.** Identifies recurring multi-segment
   correspondences and memorises those whose promotion improves
   BIC.
5. **Tonal aggregation.** Collects source-to-target tone counts
   from the converged alignments.
6. **Cross-dimensional discovery.** Searches for rules where a
   segmental feature on the source conditions a suprasegmental
   value on the target. Commits by BIC.
7. **Long-range context discovery.** A second split loop for
   conditioning environments that look beyond the immediate
   neighbours (distance-bounded, existential, and
   syllable-structural predicates).

The public entry point is :func:`train_model`, which accepts
either a pair corpus (for backward compatibility with two-lect
use) or a multi-lect ``list[CognateSet]``. A convenience
function :func:`align_corpus` applies a trained model to a
corpus.
"""

import warnings
from collections.abc import Sequence

from regulae.model import (
    CognateSet,
    LearnedModel,
    MultiLectModel,
)
from regulae.search import DEFAULT_MAX_CHUNK_SIZE, align_forms
from regulae.types import (
    Alignment,
    Form,
)

from regulae._em import (
    _displacement_aggregation,
    _initial_model,
    _replace_segment_table,
    _segment_em,
    _update_segment_table,
)
from regulae._chunks import _chunk_promotion
from regulae._discovery import (
    _context_discovery,
    _long_range_discovery,
    _tonal_aggregation,
)
from regulae._cross_dimensional import (
    _CROSS_DIM_MIN_RULE_CONFIDENCE,
    _CROSS_DIM_MIN_RULE_COUNT,
    _cross_dimensional_discovery,
    _dedup_dual_framings,
    _rule_signature,
)
from regulae._reconciliation import (
    _commit_multi_lect_long_range_splits_for_pivot,
    _lift_cross_dimensional_rules,
    _train_multi_lect,
)


# ----- defaults ------------------------------------------------------------

DEFAULT_TEMPERATURE: float = 1.0
DEFAULT_CONCENTRATION: float = 5.0
DEFAULT_MAX_ITER: int = 30
DEFAULT_CONVERGENCE_EPS: float = 1e-4
DEFAULT_SEGMENT_WEIGHT: float = 0.7
DEFAULT_DISPLACEMENT_WEIGHT: float = 0.3
DEFAULT_TONE_WEIGHT: float = 1.0


# ----- public entry points -------------------------------------------------


def train_model(
    corpus: Sequence[tuple[Form, Form]] | Sequence[CognateSet],
    *,
    feature_system: str = "descriptive",
    max_chunk_size: int = DEFAULT_MAX_CHUNK_SIZE,
    temperature: float = DEFAULT_TEMPERATURE,
    concentration: float = DEFAULT_CONCENTRATION,
    max_iter: int = DEFAULT_MAX_ITER,
    convergence_eps: float = DEFAULT_CONVERGENCE_EPS,
    segment_weight: float = DEFAULT_SEGMENT_WEIGHT,
    displacement_weight: float = DEFAULT_DISPLACEMENT_WEIGHT,
    tone_weight: float = DEFAULT_TONE_WEIGHT,
    multi_lect_bic_correction: bool = True,
    multi_lect_min_commit_scale: float = 0.5,
) -> LearnedModel | MultiLectModel:
    """Train a layered alignment model on a training corpus.

    Polymorphic over two input shapes:

    * ``list[tuple[Form, Form]]`` — legacy pairwise input. Returns a
      :class:`LearnedModel`. The pair form is deprecated; new code
      should construct :class:`CognateSet` objects.
    * ``list[CognateSet]`` — the canonical multi-lect input.
      Returns a :class:`MultiLectModel` containing per-pair learned
      models plus reconciled multi-lect correspondence classes.

    Empty input is handled as the legacy path (returns an empty
    :class:`LearnedModel`) because that's what existing call sites
    expect.

    Multi-lect-only knobs (ignored for pair input):

    * ``multi_lect_bic_correction``: add an AICc-style small-sample
      correction ``2/max(n-1, 1)`` to the class-discovery BIC
      penalty. Default True — without it, tiny pivot buckets pass
      BIC too easily.
    * ``multi_lect_min_commit_scale``: scales the adaptive floor
      for per-sister-tuple emissions in the class-discovery loop.
      The floor is ``max(2, ceil(scale * log2(n+1)))`` where
      ``n`` is the pivot's observation count. Default ``0.5``
      (gentle sub-linear scaling, tuned on multi-lect real-data
      experiments); set to ``0.0`` to disable the adaptive floor
      entirely (structural minimum of 2); set higher to tighten
      further.

    Training is deterministic given the corpus and hyperparameters.
    """
    if not corpus:
        return LearnedModel.empty(
            feature_system=feature_system,
            temperature=temperature,
            concentration=concentration,
        )

    if isinstance(corpus[0], CognateSet):
        _validate_cognate_sets(corpus)  # type: ignore[arg-type]
        return _train_multi_lect(
            corpus,  # type: ignore[arg-type]
            feature_system=feature_system,
            max_chunk_size=max_chunk_size,
            temperature=temperature,
            concentration=concentration,
            max_iter=max_iter,
            convergence_eps=convergence_eps,
            segment_weight=segment_weight,
            displacement_weight=displacement_weight,
            tone_weight=tone_weight,
            multi_lect_bic_correction=multi_lect_bic_correction,
            multi_lect_min_commit_scale=multi_lect_min_commit_scale,
        )

    warnings.warn(
        "Passing a list of (Form, Form) pairs to train_model is "
        "deprecated; use cognate_sets_from_pairs(pairs, lect_ids) to "
        "convert to a list[CognateSet] and pass that instead. The "
        "pair-input branch may be removed in a future release.",
        DeprecationWarning,
        stacklevel=2,
    )
    return _train_pairwise_legacy(
        corpus,  # type: ignore[arg-type]
        feature_system=feature_system,
        max_chunk_size=max_chunk_size,
        temperature=temperature,
        concentration=concentration,
        max_iter=max_iter,
        convergence_eps=convergence_eps,
        segment_weight=segment_weight,
        displacement_weight=displacement_weight,
        tone_weight=tone_weight,
    )


def _validate_cognate_sets(corpus: Sequence[CognateSet]) -> None:
    """Check cognate sets for common structural errors before training."""
    all_lects: set[str] = set()
    for cs in corpus:
        all_lects.update(cs.forms)
    n_lects = len(all_lects)

    for cs in corpus:
        if n_lects >= 2 and len(cs.forms) < 2:
            raise ValueError(
                f"CognateSet {cs.cognate_id!r} has {len(cs.forms)} form(s); "
                f"at least 2 are required for multi-lect training."
            )
        for lect_id, form in cs.forms.items():
            if not form.segments:
                raise ValueError(
                    f"CognateSet {cs.cognate_id!r}, lect {lect_id!r}: "
                    f"form has no segments."
                )
        if not 0.0 <= cs.confidence <= 1.0:
            raise ValueError(
                f"CognateSet {cs.cognate_id!r}: confidence must be in [0, 1], "
                f"got {cs.confidence!r}."
            )


def _train_pairwise_legacy(
    corpus: Sequence[tuple[Form, Form]],
    *,
    pair_weights: Sequence[float] | None = None,
    feature_system: str,
    max_chunk_size: int,
    temperature: float,
    concentration: float,
    max_iter: int,
    convergence_eps: float,
    segment_weight: float,
    displacement_weight: float,
    tone_weight: float,
) -> LearnedModel:
    """Internal: the pair-based training pipeline."""
    if pair_weights is None:
        pair_weights = [1.0] * len(corpus)
    if len(pair_weights) != len(corpus):
        raise ValueError("pair_weights must have the same length as corpus")

    initial = _initial_model(
        corpus=corpus,
        feature_system=feature_system,
        temperature=temperature,
        concentration=concentration,
        segment_weight=segment_weight,
        displacement_weight=displacement_weight,
    )

    after_em = _segment_em(
        corpus=corpus,
        pair_weights=pair_weights,
        initial_model=initial,
        max_chunk_size=max_chunk_size,
        max_iter=max_iter,
        convergence_eps=convergence_eps,
    )

    after_displacement = _displacement_aggregation(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_em,
        max_chunk_size=max_chunk_size,
    )

    after_context = _context_discovery(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_displacement,
        max_chunk_size=max_chunk_size,
    )

    after_chunks = _chunk_promotion(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_context,
        max_chunk_size=max_chunk_size,
    )

    after_tonal = _tonal_aggregation(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_chunks,
        max_chunk_size=max_chunk_size,
    )

    after_cross_dim = _cross_dimensional_discovery(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_tonal,
    )

    after_long_range = _long_range_discovery(
        corpus=corpus,
        pair_weights=pair_weights,
        model=after_cross_dim,
        max_chunk_size=max_chunk_size,
    )

    return after_long_range


def cognate_sets_from_pairs(
    pairs: Sequence[tuple[Form, Form]],
    lect_ids: tuple[str, str] = ("src", "tgt"),
    *,
    cognate_id_prefix: str = "pair",
) -> list[CognateSet]:
    """Convert a list of pairwise form tuples to a list of CognateSets.

    Helper for code that operates on pair lists. Each
    ``(src_form, tgt_form)`` becomes a CognateSet with a synthetic
    cognate ID and the given pair of lect IDs.

    The lect IDs supplied here are the identifiers used in the
    resulting ``CognateSet.forms`` mapping and in the
    ``Form.lect_id`` of each form (both are rewritten to match).
    """
    src_lect, tgt_lect = lect_ids
    out: list[CognateSet] = []
    for i, (src, tgt) in enumerate(pairs):
        src_form = Form(
            lect_id=src_lect, segments=src.segments, syllable_breaks=src.syllable_breaks
        )
        tgt_form = Form(
            lect_id=tgt_lect, segments=tgt.segments, syllable_breaks=tgt.syllable_breaks
        )
        out.append(
            CognateSet(
                cognate_id=f"{cognate_id_prefix}.{i:05d}",
                forms={src_lect: src_form, tgt_lect: tgt_form},
            )
        )
    return out


def align_corpus(
    corpus: Sequence[tuple[Form, Form]],
    model: LearnedModel,
    *,
    max_chunk_size: int = DEFAULT_MAX_CHUNK_SIZE,
) -> list[Alignment]:
    """Align every pair in a corpus under a trained model.

    Convenience wrapper around :func:`align_forms` with the model
    parameter fixed. Returns one :class:`Alignment` per input pair, in
    the same order as the corpus.
    """
    return [
        align_forms(src, tgt, max_chunk_size=max_chunk_size, model=model)
        for src, tgt in corpus
    ]

File: training.go
package regulae

import "fmt"

// Training orchestration. The full TrainModel entry point and the multi-lect
// pipeline are assembled in the reconciliation/training slice; this file
// currently holds the shared training defaults, the internal pair-corpus
// representation, and the alignCorpus convenience used by the EM and discovery
// stages.

// Training defaults, matching the Python module-level constants.
const (
	defaultTemperature        = 1.0
	defaultConcentration      = 5.0
	defaultMaxIter            = 30
	defaultConvergenceEps     = 1e-4
	defaultSegmentWeight      = 0.7
	defaultDisplacementWeight = 0.3
	defaultToneWeight         = 1.0
)

// FormPair is the internal pairwise-corpus element: a directed (source,
// target) form pair. The Python pipeline used tuple[Form, Form].
type FormPair struct {
	Src Form
	Tgt Form
}

// alignCorpus aligns every pair in a corpus under a model, returning one
// Alignment per input pair in corpus order. Panics on unknown graphemes
// (internal use; public boundaries recover).
func alignCorpus(corpus []FormPair, model *LearnedModel, maxChunkSize int) []Alignment {
	out := make([]Alignment, len(corpus))
	for i, p := range corpus {
		out[i] = alignForms(p.Src, p.Tgt, model.FeatureSystem, maxChunkSize, model)
	}
	return out
}

// TrainOptions bundles the tunable hyperparameters for training. Use
// DefaultTrainOptions for the standard configuration and override fields as
// needed. The zero value is NOT valid — always start from DefaultTrainOptions.
type TrainOptions struct {
	FeatureSystem        string
	MaxChunkSize         int
	Temperature          float64
	Concentration        float64
	MaxIter              int
	ConvergenceEps       float64
	SegmentWeight        float64
	DisplacementWeight   float64
	ToneWeight           float64
	ChunkMinTransparency float64
	TypologicalPrior     TypologicalPrior
	BICConfig            BICConfig
	// BootstrapN > 0 replaces closed-form Wilson intervals with bootstrap
	// percentile intervals over BootstrapN resampled trainings; BootstrapSeed
	// seeds the resampler. Training time scales as (1 + BootstrapN) × base.
	BootstrapN    int
	BootstrapSeed int
}

// DefaultTrainOptions returns the standard training configuration, matching the
// Python train_model defaults.
func DefaultTrainOptions() TrainOptions {
	return TrainOptions{
		FeatureSystem:        "descriptive",
		MaxChunkSize:         DefaultMaxChunkSize,
		Temperature:          defaultTemperature,
		Concentration:        defaultConcentration,
		MaxIter:              defaultMaxIter,
		ConvergenceEps:       defaultConvergenceEps,
		SegmentWeight:        defaultSegmentWeight,
		DisplacementWeight:   defaultDisplacementWeight,
		ToneWeight:           defaultToneWeight,
		ChunkMinTransparency: 0.0,
		TypologicalPrior:     nil,
		BICConfig:            DefaultBICConfig(),
	}
}

// TrainPairwise trains a layered LearnedModel on a directed pairwise corpus.
// pairWeights (one per pair; nil = all 1.0) are confidence weights. Returns an
// *UnknownGraphemeError if a grapheme is unrecognised.
func TrainPairwise(corpus []FormPair, pairWeights []float64, opts TrainOptions) (model *LearnedModel, err error) {
	defer catchUnknownGrapheme(&err)
	if len(corpus) == 0 {
		return EmptyLearnedModel(opts.FeatureSystem, opts.Temperature, opts.Concentration), nil
	}
	model = trainPairwiseLegacy(corpus, pairWeights, opts)
	if opts.BootstrapN > 0 {
		model = bootstrapPairwiseUncertainty(model, corpus, opts.BootstrapN, opts.BootstrapSeed, opts)
	}
	return model, nil
}

// TrainModel trains a multi-lect alignment model on a corpus of cognate sets:
// per-pair learned models plus reconciled multi-lect correspondence classes.
// This is the canonical entry point. Returns an error on invalid cognate sets
// or an unrecognised grapheme.
func TrainModel(corpus []CognateSet, opts TrainOptions) (model *MultiLectModel, err error) {
	defer catchUnknownGrapheme(&err)
	if len(corpus) == 0 {
		return EmptyMultiLectModel(), nil
	}
	if err := validateCognateSets(corpus); err != nil {
		return nil, err
	}
	model = trainMultiLect(corpus, opts)
	if opts.BootstrapN > 0 {
		model = bootstrapMultiLectUncertainty(model, corpus, opts.BootstrapN, opts.BootstrapSeed, opts)
	}
	return model, nil
}

// validateCognateSets checks cognate sets for common structural errors.
func validateCognateSets(corpus []CognateSet) error {
	allLects := map[string]bool{}
	for _, cs := range corpus {
		for l := range cs.Forms {
			allLects[l] = true
		}
	}
	nLects := len(allLects)
	for _, cs := range corpus {
		if nLects >= 2 && len(cs.Forms) < 2 {
			return fmt.Errorf("CognateSet %q has %d form(s); at least 2 are required for multi-lect training", cs.CognateID, len(cs.Forms))
		}
		for lectID, form := range cs.Forms {
			if len(form.Segments) == 0 {
				return fmt.Errorf("CognateSet %q, lect %q: form has no segments", cs.CognateID, lectID)
			}
		}
		if cs.Confidence < 0.0 || cs.Confidence > 1.0 {
			return fmt.Errorf("CognateSet %q: confidence must be in [0, 1], got %v", cs.CognateID, cs.Confidence)
		}
	}
	return nil
}

// CognateSetsFromPairs converts a list of pairwise form tuples to cognate sets,
// rewriting each form's lect id to match lectIDs and recording the (src, tgt)
// ordering in FormsOrder so the canonical lect direction is preserved.
func CognateSetsFromPairs(pairs []FormPair, lectIDs [2]string, cognateIDPrefix string) []CognateSet {
	if cognateIDPrefix == "" {
		cognateIDPrefix = "pair"
	}
	srcLect, tgtLect := lectIDs[0], lectIDs[1]
	out := make([]CognateSet, 0, len(pairs))
	for i, p := range pairs {
		srcForm := Form{LectID: srcLect, Segments: p.Src.Segments, SyllableBreaks: p.Src.SyllableBreaks, MorphemeBreaks: p.Src.MorphemeBreaks}
		tgtForm := Form{LectID: tgtLect, Segments: p.Tgt.Segments, SyllableBreaks: p.Tgt.SyllableBreaks, MorphemeBreaks: p.Tgt.MorphemeBreaks}
		out = append(out, CognateSet{
			CognateID:  fmt.Sprintf("%s.%05d", cognateIDPrefix, i),
			Forms:      map[string]Form{srcLect: srcForm, tgtLect: tgtForm},
			FormsOrder: []string{srcLect, tgtLect},
			Confidence: 1.0,
		})
	}
	return out
}

// AlignCorpus aligns every pair in a corpus under a trained model, returning
// one Alignment per input pair in corpus order. Returns an *UnknownGraphemeError
// on an unrecognised grapheme.
func AlignCorpus(corpus []FormPair, model *LearnedModel, maxChunkSize int) (out []Alignment, err error) {
	defer catchUnknownGrapheme(&err)
	if maxChunkSize == 0 {
		maxChunkSize = DefaultMaxChunkSize
	}
	out = alignCorpus(corpus, model, maxChunkSize)
	return out, nil
}

// trainPairwiseLegacy runs the staged pairwise training pipeline:
// initial prior -> segment EM -> displacement aggregation -> context discovery
// -> chunk promotion -> tonal aggregation -> cross-dimensional discovery ->
// long-range context discovery.
func trainPairwiseLegacy(corpus []FormPair, pairWeights []float64, opts TrainOptions) *LearnedModel {
	if pairWeights == nil {
		pairWeights = make([]float64, len(corpus))
		for i := range pairWeights {
			pairWeights[i] = 1.0
		}
	}

	initial := initialModel(corpus, pairWeights, opts.FeatureSystem, opts.Temperature, opts.Concentration, opts.SegmentWeight, opts.DisplacementWeight, opts.TypologicalPrior)
	afterEM := segmentEM(corpus, pairWeights, initial, opts.MaxChunkSize, opts.MaxIter, opts.ConvergenceEps)
	afterDisplacement := displacementAggregation(corpus, pairWeights, afterEM, opts.MaxChunkSize)
	afterContext := contextDiscovery(corpus, pairWeights, afterDisplacement, opts.MaxChunkSize, opts.BICConfig)
	afterChunks := chunkPromotion(corpus, pairWeights, afterContext, opts.MaxChunkSize, opts.ChunkMinTransparency, opts.BICConfig)
	afterTonal := tonalAggregation(corpus, pairWeights, afterChunks, opts.MaxChunkSize)
	afterCrossDim := crossDimensionalDiscovery(corpus, pairWeights, afterTonal, 0, opts.BICConfig)
	afterLongRange := longRangeDiscovery(corpus, pairWeights, afterCrossDim, opts.MaxChunkSize, opts.BICConfig)
	return afterLongRange
}
